test/mocks/mock_server.js

```diff
@@ -0,0 +1,250 @@
+  .option('-t, --total-updates <n>',
+    `Total number of updates available in the feed (default: ${defaults.totalUpdates})`,
+    Number, defaults.totalUpdates)
+  .parse(process.argv);
+
+const cfg = {
+  abortEvery: program.abortEvery,
+  heartbeat: program.heartbeat,
+  lastSeqEvery: program.lastSeqEvery,
+  maxDelay: program.maxDelay,
+  minDelay: program.minDelay,
+  port: program.port,
+  probOfDelay: program.delayProbability,
+  probOf500: program.errorProbability,
+  since: 0,
+  totalUpdates: program.totalUpdates
+};
+
+console.log('Mock server configuration:');
+console.log(JSON.stringify(cfg, null, 4));
+
+// utils
+
+function getRandomArbitrary(min, max) {
+  return Math.random() * (max - min) + min;
+}
+
+function createSleepPromise(timeout) {
+  return new Promise(function(resolve) {
+    setTimeout(resolve, timeout);
+  });
+}
+
+function sleep(timeout) {
+  function promiseFunction(value) {
+    return createSleepPromise(timeout).then(function() {
+      return value;
+    });
+  }
+  promiseFunction.then = function() {
+    var sleepPromise = createSleepPromise(timeout);
+    return sleepPromise.then.apply(sleepPromise, arguments);
+  };
+  promiseFunction.catch = Promise.resolve().catch;
+  return promiseFunction;
+}
+
+// response
+
+class Response {
+  constructor(request, response) {
+    this._cfg = Object.assign({}, cfg);
+
+    this.response = response;
+    this.request = request;
+
+    this.query = this._parseQuery();
+
+    if (this.query.pathname.endsWith('_changes')) {
+      this.requestType = 'changes';
+    } else if (this.query.pathname.endsWith('_db_updates')) {
+      this.requestType = 'db_updates';
+    } else {
+      this.requestType = 'other';
+    }
+  }
+
+  _parseQuery() {
+    var parsed = url.parse(this.request.url);
+    var query = parsed.query ? querystring.parse(parsed.query) : {};
+
+    console.log(`Received request: ${url.format(parsed)}`);
+
+    var pathname = parsed.pathname;
+    if (pathname.endsWith('/')) {
+      pathname = pathname.slice(0, -1); // remove trailing slash
+    }
+    query.pathname = pathname;
+
+    // parse query arguments
+    if (query.heartbeat) {
+      this._cfg.heartbeat = parseInt(query.heartbeat, 10);
+    }
+    if (query.since) {
+      this._cfg.since = parseInt(query.since.split('-')[0], 10);
+    }
+
+    return query;
+  }
+
+  send() {
+    var self = this;
+
+    // set CT header
+    self.response.setHeader('Content-Type', ['application/json', 'charset=UTF-8']);
+
+    if (self.requestType === 'other') {
+      // return 200 OK
+      self.response.writeHead(200);
+      self.response.end('{"db_name":"foobar","instance_start_time":123,"couchdb":"Welcome"}\n');
+      return;
+    }
+
+    // set TE header
+    self.response.setHeader('Transfer-Encoding', 'chunked');
+
+    var updateCount = self._cfg.since; // start seq
+
+    if (Math.random() <= self._cfg.probOf500) {
+      // send 500
+      console.log('Returning 500!');
```

Comment:
Should these console logs use debug or some other logger?

Reply:
This just writes to stdout. The parent process than spawns this mock server is capturing all stdout/stderr and using `debug` to output to the console. I think if I used `debug` here too then we'd see two timestamps.